File: Java/SnomedQuery/SnomedQuery/Model/SnomedModelManager.java
package SnomedQuery.Model;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import org.joda.time.DateTime;

/**
 * <p>
 * SnomedModelManager class, handles data parsing and queries.
 * </p>
 *
 * @author Travis Lukach
 *
 */
public class SnomedModelManager {

	/**
	 * <p>
	 * The serialized character set is UTF-8. Mainly used when creating new
	 * String objects.
	 * </p>
	 */
	public final String _charSet = "UTF-8";

	/**
	 * <p>
	 * Space String for validating deseralized String objects.
	 * </p>
	 */
	public final String _space = " ";

	/**
	 * <p>
	 * The null character for UTF-8, used for validating deserialized Strings.
	 * </p>
	 */
	public final char _UTF8Null = '\u00bd';

	/**
	 * <p>
	 * java.nio.ByteBuffer wraps a byte array for serializing
	 * <b>SnomedQueryConcept</b> concepts.
	 * </p>
	 */
	private ByteArrayOutputStream binaryWriter;

	/**
	 * <p>
	 * java.nio.ByteBuffer wraps a byte array for serailizing
	 * <b>SnomedQueryConcept</b> relationships.
	 * </p>
	 */
	private ByteArrayOutputStream relationshipWriter;

	/**
	 * <p>
	 * java.util.HashMap for holding a unique set of <b>SnomedQueryConcept</b>
	 * objects.
	 * </p>
	 */
	private HashMap<Long, SnomedQueryConcept> snomedConcepts = new HashMap<Long, SnomedQueryConcept>();

	/**
	 * <p>
	 * Base directory for data files when deserializing.
	 * </p>
	 */
	private String baseDir;

	/**
	 * <p>
	 * java.nio.ByteBuffer wraps a byte array for reading and constructing
	 * <b>SnomedQueryConcept</b> object.
	 * </p>
	 */
	private ByteBuffer binaryReader;

	/**
	 * <p>
	 * Constructs a Path to the RawData directory.
	 * </p>
	 *
	 * @return The RawData sub-directory of baseDir.
	 */
	public String getRawDataDir() {
		return this.baseDir + "\\" + "RawData";
	}

	/**
	 * <p>
	 * Constructs a Path to the ParsedData directory.
	 * </p>
	 *
	 * @return The ParsedData sub-directory of baseDir.
	 */
	public String getParsedRecordsDir() {
		return this.baseDir + "\\" + "ParsedRecords";
	}

	/**
	 * <p>
	 * Sets the <b>baseDir</b> object with <i>baseDirParam</i> parameter.
	 * </p>
	 *
	 * @param baseDirParam
	 *            Sets the baseDir value.
	 */
	public SnomedModelManager(String baseDirParam) {
		this.baseDir = baseDirParam;
	}

	/**
	 * <p>
	 * Loads the SNOMED Concepts into memory.
	 * </p>
	 */
	public void loadRecords() {
		this.deserialize();
	}

	/**
	 * <p>
	 * Deserialize all records serialized in C#.
	 * </p>
	 */
	public void deserialize() {
		LinkedList<SnomedQueryConcept> list = new LinkedList<>();
		try {
			FileInputStream dataReader = new FileInputStream(
					new File(this.getParsedRecordsDir() + "\\"
							+ "SnomedQueryConcepts.Data.ser"));
			this.binaryReader = ByteBuffer
					.wrap(new byte[(int) dataReader.getChannel().size()]);
			dataReader.getChannel().read(this.binaryReader);
			this.binaryReader.position(0);
			this.binaryReader.order(ByteOrder.LITTLE_ENDIAN);
			int conceptCount = this.binaryReader.getInt();
			for (int i = 0; i < conceptCount; i++) {
				SnomedQueryConcept concept = this.deserializeConceptData();
				this.add(concept);
				list.addLast(concept);
			}
			dataReader.close();
		} catch (IOException e) {
			e.printStackTrace();
		} catch (Exception e) {
			e.printStackTrace();
		}

		try {
			FileInputStream dataReader = new FileInputStream(
					new File(this.getParsedRecordsDir() + "\\"
							+ "SnomedQueryConcepts.IsARelationships.ser"));
			byte[] fileBytes = new byte[(int) dataReader.getChannel().size()];
			dataReader.read(fileBytes);
			this.binaryReader = ByteBuffer.wrap(fileBytes)
					.order(ByteOrder.LITTLE_ENDIAN);
			for (SnomedQueryConcept concept : list) {
				this.deserializeConceptRealtionship(concept);
			}
			dataReader.close();
		} catch (IOException e) {
			e.printStackTrace();

		} catch (Exception e) {
			e.printStackTrace();
		}

	}

	/**
	 * <p>
	 * Add SNOMED query concept to model manager.
	 * </p>
	 *
	 * @param concept
	 *            The concept that will be inserted into <b>snomedConcepts</b>.
	 * @throws Exception
	 *             General exception, custom message.
	 */
	public void add(SnomedQueryConcept concept) throws Exception {
		if (this.snomedConcepts.putIfAbsent(concept.getConceptId(),
				concept) != null)
			System.err.println(
					String.format("Error adding concept %s to dictionary.`",
							concept.getConceptId()));
	}

	SnomedQueryConcept deserializeConceptData() {
		long conceptId = this.binaryReader.getLong();
		String conceptFullyQualifiedName = this.getString();
		String[] conceptSynonyms = this.getStringArray();
		String conceptModule = this.getString();
		String conceptDefinitionStatus = this.getString();
		DateTime conceptEffectiveTime = this.getDateTime();

		SnomedQueryConcept concept = new SnomedQueryConcept(conceptId,
				conceptFullyQualifiedName, conceptSynonyms, conceptModule,
				conceptDefinitionStatus, conceptEffectiveTime);
		concept.setIsAParents(
				new SnomedQueryConcept[this.binaryReader.getInt()]);
		concept.setIsAChildren(
				new SnomedQueryConcept[this.binaryReader.getInt()]);
		return concept;
	}

	/**
	 * <p>
	 * Gets the length of the char[] to be deserialized first, then deserializes
	 * the bytes before validating them.
	 * </p>
	 *
	 * @return Parsed and validated string created from byte[].
	 */
	String getString() {
		String retVal;
		int length = this.binaryReader.getInt();
		char[] readChars = new char[length];
		for (int i = 0; i < length; i++) {
			readChars[i] = this.binaryReader.order(ByteOrder.LITTLE_ENDIAN)
					.getChar();
		}
		retVal = new String(readChars);
		return retVal;

	}

	/**
	 * <p>
	 * Constructs a String array by reading th length of the String array first
	 * out of <b>binaryReader</b> before.
	 * </p>
	 *
	 * @return Deserialized String array.
	 */
	String[] getStringArray() {
		int length = this.binaryReader.getInt();
		String[] retVal = new String[length];
		for (int i = 0; i < length; i++) {
			retVal[i] = getString();
		}
		return retVal;
	}

	/**
	 * <p>
	 * constructs a JodaTime DateTime object out of <b>binaryReader</b>.
	 * </p>
	 *
	 * @return A new DateTime object.
	 */
	DateTime getDateTime() {
		return new DateTime(this.binaryReader.getLong());
	}

	/**
	 * <p>
	 * Link up parent/child relationships
	 * </p>
	 *
	 * @param concept
	 *            Concept to have relationships linked to it.
	 */
	void deserializeConceptRealtionship(SnomedQueryConcept concept) {
		SnomedQueryConcept[] childConcepts = concept.getIsAChildren();
		for (int i = 0; i < childConcepts.length; i++) {
			long childConceptId = this.binaryReader.getLong();
			SnomedQueryConcept child = this.getConceptById(childConceptId);
			SnomedQueryConcept[] childParents = child.getIsAParents();
			childParents[this.binaryReader.getInt()] = concept;
			child.setIsAParents(childParents);
			childConcepts[i] = child;
			concept.setIsAChildren(childConcepts);
		}
	}

	/**
	 * <p>
	 * Serialize concepts and relationships in memory to binary
	 * </p>
	 */
	public void serialize() {
		this.binaryWriter = new ByteArrayOutputStream();
		this.relationshipWriter = new ByteArrayOutputStream();
		try {
			FileOutputStream dataOutput = new FileOutputStream(
					new File(this.getParsedRecordsDir() + "\\"
							+ "SnomedQueryConcepts.Data.ser"));
			FileOutputStream relOutput = new FileOutputStream(
					new File(this.getParsedRecordsDir() + "\\"
							+ "SnomedQueryConcepts.IsARelationships.ser"));
			this.binaryWriter = this.write(this.binaryWriter,
					this.snomedConcepts.size());
			for (SnomedQueryConcept concept : this.snomedConcepts.values()) {
				this.serializeConcept(concept);
			}

			dataOutput.write(this.binaryWriter.toByteArray());
			relOutput.write(this.relationshipWriter.toByteArray());

			dataOutput.flush();
			dataOutput.close();

			relOutput.flush();
			relOutput.close();

		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	/**
	 * <p>
	 * Serializes the concept and relationships.
	 * </p>
	 * 
	 * @param concept
	 * @throws Exception
	 */
	void serializeConcept(SnomedQueryConcept concept) throws Exception {
		if (concept.getConceptId() == 45614000)
			System.out.println("");
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptId());
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptFullyQualifiedName());
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptSynonyms());
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptModule());
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptDefinitionStatus());
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getConceptEffectiveTime().getMillis());

		this.binaryWriter = this.write(this.binaryWriter,
				concept.getIsAParents().length);
		this.binaryWriter = this.write(this.binaryWriter,
				concept.getIsAChildren().length);

		SnomedQueryConcept[] childArray = concept.getIsAChildren();
		for (int i = 0; i < childArray.length; i++) {
			SnomedQueryConcept child = childArray[i];
			this.relationshipWriter = this.write(this.relationshipWriter,
					child.getConceptId());
			int reverseIndex = -1;
			SnomedQueryConcept[] parentArray = child.getIsAParents();
			for (int j = 0; j < parentArray.length; j++) {
				if (parentArray[j].getConceptId() == concept.getConceptId()) {
					reverseIndex = j;
					break;
				}
			}

			if (reverseIndex == -1) {
				throw new Exception("Reverse snomed index not found");
			}
			this.relationshipWriter = this.write(this.relationshipWriter,
					reverseIndex);
		}
	}

	/**
	 * <p>
	 * Serializes an integer value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer, int value)
			throws IOException {
		buffer.write(ByteBuffer.allocate(Integer.BYTES)
				.order(ByteOrder.LITTLE_ENDIAN).putInt(value).array());
		return buffer;
	}

	/**
	 * <p>
	 * Serializes a long value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer, long value)
			throws IOException {
		buffer.write(ByteBuffer.allocate(Long.BYTES)
				.order(ByteOrder.LITTLE_ENDIAN).putLong(value).array());
		return buffer;
	}

	/**
	 * <p>
	 * Serializes a long value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer,
			DateTime value) throws IOException {
		buffer.write(
				ByteBuffer.allocate(Long.BYTES).order(ByteOrder.LITTLE_ENDIAN)
						.putLong(value.getMillis()).array());
		return buffer;
	}

	/**
	 * <p>
	 * Serializes a String value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer,
			String value) throws IOException {
		buffer.write(ByteBuffer.allocate(Integer.BYTES)
				.order(ByteOrder.LITTLE_ENDIAN).putInt(value.length()).array());
		for (int i = 0; i < value.length(); i++) {
			buffer.write(ByteBuffer.allocate(Character.BYTES)
					.order(ByteOrder.LITTLE_ENDIAN)
					.putChar(value.toCharArray()[i]).array());
		}
		return buffer;
	}

	/**
	 * <p>
	 * Serializes a long array value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer,
			long[] value) throws IOException {
		buffer.write(ByteBuffer.allocate(Integer.BYTES)
				.order(ByteOrder.LITTLE_ENDIAN).putInt(value.length).array());
		for (long l : value) {
			buffer.write(ByteBuffer.allocate(Long.BYTES)
					.order(ByteOrder.LITTLE_ENDIAN).putLong(l).array());
		}
		return buffer;
	}

	/**
	 * <p>
	 * Serializes a string array value.
	 * </p>
	 * 
	 * @param buffer
	 * @param value
	 * @return
	 * @throws IOException
	 */
	public ByteArrayOutputStream write(ByteArrayOutputStream buffer,
			String[] value) throws IOException {
		buffer.write(ByteBuffer.allocate(Integer.BYTES)
				.order(ByteOrder.LITTLE_ENDIAN).putInt(value.length).array());
		for (int i = 0; i < value.length; i++) {
			buffer = this.write(buffer, value[i]);
		}
		return buffer;
	}

	/**
	 * <p>
	 * Get concept by its id. Return null if concept not found.
	 * </p>
	 *
	 * @param id
	 *            searched
	 * @return the queried id. Null if not found.
	 */
	public SnomedQueryConcept getConceptById(long id) {
		SnomedQueryConcept retVal;
		retVal = this.snomedConcepts.get(id);
		return retVal;
	}

	/**
	 * <p>
	 * Return true if concept with child id is a child of parent.
	 * </p>
	 *
	 * @param visitedItems
	 *            Persisted list of items that have already been viewed.
	 * @param parent
	 *            SnomedQueryConcept of which children is about to be parsed.
	 * @param childId
	 *            Target child Id that is being searched.
	 * @return True if there is a child object with child id.
	 */
	public boolean isChild(HashMap<Long, SnomedQueryConcept> visitedItems,
			SnomedQueryConcept parent, long childId) {
		SnomedQueryConcept[] children = parent.getIsAChildren();
		for (SnomedQueryConcept child : children) {
			if (!visitedItems.containsKey(child.getConceptId())) {
				visitedItems.put(child.getConceptId(), child);
				if (child.getConceptId() == childId)
					return true;
				if (this.isChild(visitedItems, child, childId) == true)
					return true;
			}
		}
		return false;
	}

	/**
	 * <p>
	 * Return true if parent is an ancestor of child.
	 * </p>
	 *
	 * @param parentId
	 *            Parent Id.
	 * @param childId
	 *            Child Id.
	 * @return True if there is a child ancestor.
	 */
	public boolean isChild(long parentId, long childId) {
		SnomedQueryConcept parent = this.getConceptById(parentId);
		HashMap<Long, SnomedQueryConcept> visitedItems = new HashMap<>();
		return this.isChild(visitedItems, parent, childId);
	}

	/**
	 * Collect children of parent and continue visitor pattern onward with
	 * children.
	 *
	 * @param visitedItems
	 *            Items already visited.
	 * @param parent
	 *            Parent concept.
	 * @return the map of visited items.
	 */
	public HashMap<Long, SnomedQueryConcept> findDecendents(
			HashMap<Long, SnomedQueryConcept> visitedItems,
			SnomedQueryConcept parent) {
		for (SnomedQueryConcept child : parent.getIsAChildren()) {
			if (!visitedItems.containsKey(child.getConceptId())) {
				visitedItems.put(child.getConceptId(), child);
				this.findDecendents(visitedItems, child);
			}
		}
		return visitedItems;
	}

	/**
	 * <p>
	 * Collect child IDs of parent and continue visitor pattern onward with
	 * children.
	 * </p>
	 *
	 * @param conceptId
	 *            the queried Id.
	 * @return The array of child concepts.
	 */
	public SnomedQueryConcept[] findDescendants(long conceptId) {
		SnomedQueryConcept concept = this.getConceptById(conceptId);
		HashMap<Long, SnomedQueryConcept> visitedItems = new HashMap<>();
		visitedItems = this.findDecendents(visitedItems, concept);
		SnomedQueryConcept[] retVal = new SnomedQueryConcept[visitedItems
				.size()];
		int i = 0;
		for (SnomedQueryConcept qc : visitedItems.values()) {
			retVal[i] = qc;
			i++;
		}
		return retVal;

	}

	/**
	 * <p>
	 * Recursive Method for gathering all parents, and further up the line.
	 * </p>
	 *
	 * @param visitedItems
	 *            hashMap of gathered items.
	 * @param child
	 *            current child looking for parents.
	 * @return the collected HashMap of objects.
	 */
	public HashMap<Long, SnomedQueryConcept> findAncestors(
			HashMap<Long, SnomedQueryConcept> visitedItems,
			SnomedQueryConcept child) {
		for (SnomedQueryConcept parent : child.getIsAParents()) {
			if (!visitedItems.containsKey(parent.getConceptId())) {
				visitedItems.put(parent.getConceptId(), parent);
				this.findAncestors(visitedItems, parent);
			}
		}
		return visitedItems;
	}

	/**
	 * <p>
	 * The start of finding all the parent Id's from target Id.
	 * </p>
	 *
	 * @param conceptId
	 *            the targetId that is to be queried.
	 * @return All the parent objects in an array.
	 */
	public SnomedQueryConcept[] findAncestors(long conceptId) {
		SnomedQueryConcept concept = this.getConceptById(conceptId);
		HashMap<Long, SnomedQueryConcept> visitedItems = new HashMap<>();
		visitedItems = this.findAncestors(visitedItems, concept);
		SnomedQueryConcept[] retVal = new SnomedQueryConcept[visitedItems
				.size()];
		int i = 0;
		for (SnomedQueryConcept qc : visitedItems.values()) {
			retVal[i] = qc;
			i++;
		}
		return retVal;
	}

	/**
	 * <p>
	 * Outputs a Tab delimeted closure table file to the directory of choice.
	 * </p>
	 * 
	 * @param ouputFile
	 */
	public void createClosureTable(String ouputFile) {
		List<Long[]> outputSet = new ArrayList<>();
		Long currentSource;
		Long currentDestination;
		File outputFile = new File(ouputFile);
		try {
			FileWriter fWriter = new FileWriter(outputFile);
			PrintWriter pWriter = new PrintWriter(fWriter);
			for (SnomedQueryConcept conceptSource : this.snomedConcepts
					.values()) {
				currentSource = conceptSource.getConceptId();
				SnomedQueryConcept[] conceptArray = this
						.findDescendants(conceptSource.getConceptId());
				for (SnomedQueryConcept conceptDestination : conceptArray) {
					currentDestination = conceptDestination.getConceptId();
					outputSet
							.add(new Long[]{currentSource, currentDestination});
				}
			}
			for (Long[] pair : outputSet) {
				pWriter.write(String.format("%s\t%s\n", pair[0], pair[1]));
			}
			pWriter.close();

			fWriter.close();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: C#/Projects/SnomedQuery.Model/SnomedModelManager.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.IO;
using System.Runtime;
using System.Threading;
using System.Threading.Tasks;

namespace SnomedQuery.Model
{
  /// <summary>
  /// Snomed Model Manager class.
  /// This class contains all the data that relates to the Snomed
  /// concepts. It loads preparsed data and allows specific
  /// queries against that data.
  /// </summary>
  public partial class SnomedModelManager
  {
    /// <summary>
    /// Dictionary of all concepts ordered by their snomed id.
    /// </summary>
    ConcurrentDictionary<Int64, SnomedQueryConcept> snomedConcepts = new ConcurrentDictionary<long, SnomedQueryConcept>();

    /// <summary>
    /// Base directory for data files
    /// </summary>
    String baseDir;

    /// <summary>
    /// Location of common raw files cache directory.
    /// </summary>
    public String RawDataDir
    {
      get { return Path.Combine(this.baseDir, "RawData"); }
    }

    /// <summary>
    /// Location of common parsed records cache directory.
    /// </summary>
    public String ParsedRecordsDir
    {
      get { return Path.Combine(this.baseDir, "ParsedRecords"); }
    }

    /// <summary>
    /// Constructor.
    /// </summary>
    /// <param name="baseDirParam">
    /// Base directory for data files
    /// The layout of the data MUST be
    /// {baseDirParam}
    ///   RawData (only used in parse)
    ///   ParsedRecords (preparsed records stored here)
    /// </param>
    public SnomedModelManager(String baseDirParam)
    {
      this.baseDir = baseDirParam;
    }

    /// <summary>
    /// Load records asynchronously
    /// </summary>
    /// <returns></returns>
    public void LoadRecords()
    {
      this.Deserialize();
    }

    /// <summary>
    /// Serialize all records created in this project.
    /// </summary>
    public void Serialize()
    {
      using (Stream dataStream = new FileStream(Path.Combine(this.ParsedRecordsDir, "SnomedQueryConcepts.Data.ser"),
                                                FileMode.Create,
                                                FileAccess.Write,
                                                FileShare.None),
            relationshipStream = new FileStream(Path.Combine(this.ParsedRecordsDir, "SnomedQueryConcepts.IsARelationships.ser"),
                                                FileMode.Create,
                                                FileAccess.Write,
                                                FileShare.None)
                                                )
      {
        using (BinaryWriter dataWriter = new BinaryWriter(dataStream),
              relationshipWriter = new BinaryWriter(relationshipStream))
        {
          this.Write(dataWriter, this.snomedConcepts.Count);      // write number of concepts
          // Write data part of each concept.
          foreach (SnomedQueryConcept snomedConcept in this.snomedConcepts.Values)
            this.SerializeConcept(dataWriter, relationshipWriter, snomedConcept);

          dataWriter.Flush();
          dataWriter.Close();
        }
        dataStream.Close();
      }
    }

    /// <summary>
    /// Serialize the data part of the snomed concept instance.
    /// </summary>
    /// <param name="dataWriter">Write conept data to this stream</param>
    /// <param name="relationshipWriter">Write IsA relationships to this stream</param>
    /// <param name="snomedConcept"></param>
    void SerializeConcept(BinaryWriter dataWriter,
      BinaryWriter relationshipWriter,
      SnomedQueryConcept snomedConcept)
    {
      this.Write(dataWriter, snomedConcept.ConceptId);
      this.Write(dataWriter, snomedConcept.ConceptFullyQualifiedName);
      this.Write(dataWriter, snomedConcept.ConceptSynonymns);
      this.Write(dataWriter, snomedConcept.ConceptModule);
      this.Write(dataWriter, snomedConcept.ConceptDefinitionStatus);
      this.Write(dataWriter, snomedConcept.ConceptEffectiveTime.Ticks);

      // Write out length of parent/child relationship arrays.
      this.Write(dataWriter, snomedConcept.IsAParents.Length);
      this.Write(dataWriter, snomedConcept.IsAChildren.Length);

      /*
       * This is a bit obtuse.
       * To avoid unnecessary dictionary lookups, we serialize the isa data as follows.
       * a) Only isa children are serialized. The parent fields are derived from the
       *    children values.
       * b) The children id are serialized, and the offset in the childs parent relationship
       *    array is serialized. This makes it possible to deserialice quickly, but serialization
       *    is very much slower.
       */
      // Output number of values in child list.
      for (Int32 i = 0; i < snomedConcept.IsAChildren.Length; i++)
      {
        SnomedQueryConcept child = snomedConcept.IsAChildren[i];
        this.Write(relationshipWriter, child.ConceptId);
        Int32 reverseIndex = -1;
        for (Int32 j = 0; j < child.IsAParents.Length; j++)
        {
          if (child.IsAParents[j] == snomedConcept)
          {
            reverseIndex = j;
            break;
          }
        }

        // If reverseIndex < 1, then the child record did not have a corresponding
        // link to the parent. This is an error.
        if (reverseIndex == -1)
          throw new ApplicationException("Reverse snomed index not found");
        this.Write(relationshipWriter, reverseIndex);
      }
    }

    /// <summary>
    /// Deserialize all records serialized in Serialize().
    /// </summary>
    public void Deserialize()
    {
      // Keep list of concepts read in, in the SAME ORDER they are read in.
      // This is used later to link up relationships.
      LinkedList<SnomedQueryConcept> list = new LinkedList<SnomedQueryConcept>();

      // Read in conept data. This does not read in relationships, though the
      // relationship arrays are resized properly.
      String dataPath = Path.Combine(this.ParsedRecordsDir, "SnomedQueryConcepts.Data.ser");
      dataPath = Path.GetFullPath(dataPath);
      using (Stream inStream = new FileStream(dataPath, FileMode.Open, FileAccess.Read, FileShare.Read))
      {
        using (BinaryReader dataReader = new BinaryReader(inStream))
        {
          Int32 conceptCount = this.ReadInt32(dataReader);
          for (Int32 i = 0; i < conceptCount; i++)
          {
            SnomedQueryConcept concept = this.DeserializeConceptData(dataReader);
            this.Add(concept);
            list.AddLast(concept);
          }
        }
      }

      // Read in relationship and link concepts togother. This must be done after all concepts
      // have been created so we can access the parent and child objects (cant point to object
      // that doesn't exist yet...)
      using (Stream inStream = new FileStream(
                             Path.Combine(this.ParsedRecordsDir, "SnomedQueryConcepts.IsARelationships.ser"),
                             FileMode.Open,
                             FileAccess.Read,
                             FileShare.Read))
      {
        using (BinaryReader relationshipReader = new BinaryReader(inStream))
        {
          foreach (SnomedQueryConcept concept in list)
          {
            this.DeserializeConceptRelationships(relationshipReader, concept);
          }
        }
      }
    }


    /// <summary>
    /// Add snomed query concept to model manager.
    /// </summary>
    /// <param name="concept"></param>
    public void Add(SnomedQueryConcept concept)
    {
      if (this.snomedConcepts.TryAdd(concept.ConceptId, concept) == false)
        throw new ApplicationException($"Error adding concept {concept.ConceptId} to dictionary.`");
    }

    /// <summary>
    /// Return snomed concept record with its data initialized from the serial input file.
    /// The is a links are not set yet.
    /// </summary>
    /// <param name="binaryReader"></param>
    /// <returns>Newly created snomed query concept instance</returns>
    SnomedQueryConcept DeserializeConceptData(BinaryReader binaryReader)
    {
      SnomedQueryConcept retVal = new SnomedQueryConcept(
              this.ReadInt64(binaryReader),                // conceptId
               this.ReadString(binaryReader),               // conceptFullyQualifiedName,
               this.ReadStringArray(binaryReader),          // conceptSynonymns,
               this.ReadString(binaryReader),               // conceptModule,
               this.ReadString(binaryReader),               // conceptDefinitionStatus,
               this.ReadDateTime(binaryReader)              // conceptEffectiveTime);
             );

      // Set length of parent/child arrays. The actual values in the array will be set later.
      {
        Int32 count = this.ReadInt32(binaryReader);
        retVal.IsAParents = new SnomedQueryConcept[count];
      }

      {
        Int32 count = this.ReadInt32(binaryReader);
        retVal.IsAChildren = new SnomedQueryConcept[count];
      }
      return retVal;
    }

    /// <summary>
    /// Link up parent/child relationships
    /// </summary>
    /// <param name="relationshipReader">Read relationship data from this stream</param>
    /// <param name="concept">Concept to set child relationship data for</param>
    /// <returns>Newly created snomed query concept instance</returns>
    void DeserializeConceptRelationships(BinaryReader relationshipReader,
      SnomedQueryConcept concept)
    {
      for (Int32 i = 0; i < concept.IsAChildren.Length; i++)
      {
        Int64 childConceptId = this.ReadInt64(relationshipReader);
        SnomedQueryConcept child = this.GetConceptById(childConceptId);
        concept.IsAChildren[i] = child;
        child.IsAParents[this.ReadInt32(relationshipReader)] = concept;
      }
    }

    /// <summary>
    /// Get concept by its id. Return null if concept not found.
    /// </summary>
    /// <param name="id"></param>
    /// <returns></returns>
    public SnomedQueryConcept GetConceptById(Int64 id)
    {
      SnomedQueryConcept retVal;
      if (this.snomedConcepts.TryGetValue(id, out retVal) == false)
        return null;
      return retVal;
    }

    /// <summary>
    /// Return true if concept with child id is a child of parent.
    /// </summary>
    /// <param name="visitedItems">Items already visited. Dont go there again</param>
    /// <param name="parent">parent concept</param>
    /// <param name="childID">child concept id to search for</param>
    /// <returns></returns>
    bool IsChild(ConcurrentDictionary<Int64, SnomedQueryConcept> visitedItems,
      SnomedQueryConcept parent,
      Int64 childID)
    {
      foreach (SnomedQueryConcept child in parent.IsAChildren)
      {
        // If we have not visited this before.
        if (visitedItems.TryAdd(child.ConceptId, child) == true)
        {
          if (child.ConceptId == childID)
            return true;

          if (this.IsChild(visitedItems, child, childID) == true)
            return true;
        }
      }
      return false;
    }

    /// <summary>
    /// Return true if parent is an ancestor of child
    /// </summary>
    /// <param name="parentID"></param>
    /// <param name="childID"></param>
    /// <returns></returns>
    public bool IsChild(Int64 parentID, Int64 childID)
    {
      SnomedQueryConcept parent = this.GetConceptById(parentID);
      ConcurrentDictionary<Int64, SnomedQueryConcept> visitedItems = new ConcurrentDictionary<long, SnomedQueryConcept>();
      return this.IsChild(visitedItems, parent, childID);
    }

    /// <summary>
    /// Collect child IDs of parent and continue visitor pattern onward with children.
    /// </summary>
    /// <param name="visitedItems">Items already visited. Dont go there again</param>
    /// <param name="decendents">All IDs of children visited from public FindDecendents.</param>
    /// <param name="parent">parent concept</param>
    /// <returns></returns>
    void FindDecendents(ConcurrentDictionary<Int64, SnomedQueryConcept> visitedItems,
       SnomedQueryConcept thisConcept)
    {
      foreach (SnomedQueryConcept child in thisConcept.IsAChildren)
      {
        // If we have not visited this before.
        if (visitedItems.TryAdd(child.ConceptId, child) == true)
        {
          this.FindDecendents(visitedItems, child);
        }
      }
    }

    /// <summary>
    /// Collect child IDs of parent and continue visitor pattern onward with children.
    /// </summary>
    /// <param name="visitedItems">Items already visited. Dont go there again</param>
    /// <param name="ancestors">All IDs of parents visited from public FindAncestors.</param>
    /// <param name="parent">parent concept</param>
    /// <returns></returns>
    void FindAncestors(ConcurrentDictionary<Int64, SnomedQueryConcept> visitedItems,
       SnomedQueryConcept thisConcept)
    {
      foreach (SnomedQueryConcept parent in thisConcept.IsAParents)
      {
        // If we have not visited this before.
        if (visitedItems.TryAdd(parent.ConceptId, parent) == true)
        {
          this.FindAncestors(visitedItems, parent);
        }
      }
    }

    /// <summary>
    /// Return Ancestors concepts of parent, does not include parent.
    /// </summary>
    /// <param name="parentID"></param>
    /// <returns></returns>
    public SnomedQueryConcept[] FindAncestors(Int64 currentId)
    {
      SnomedQueryConcept currentConcept = this.GetConceptById(currentId);
      ConcurrentDictionary<Int64, SnomedQueryConcept> ancestorIDs = new ConcurrentDictionary<Int64, SnomedQueryConcept>();
      this.FindAncestors(ancestorIDs, currentConcept);

      SnomedQueryConcept[] retVal = new SnomedQueryConcept[ancestorIDs.Count];
      ancestorIDs.Values.CopyTo(retVal, 0);
      return retVal;
    }

    /// <summary>
    /// Return Descendant concepts of parent, does not include parent.
    /// </summary>
    /// <param name="parentID"></param>
    /// <returns></returns>
    public SnomedQueryConcept[] FindDecendents(Int64 currentId)
    {
      SnomedQueryConcept currentConcept = this.GetConceptById(currentId);
      ConcurrentDictionary<Int64, SnomedQueryConcept> decendentIDs = new ConcurrentDictionary<Int64, SnomedQueryConcept>();
      this.FindDecendents(decendentIDs, currentConcept);
      SnomedQueryConcept[] retVal = new SnomedQueryConcept[decendentIDs.Count];
      decendentIDs.Values.CopyTo(retVal, 0);
      return retVal;
    }


    /// <summary>
    /// Outputs a Tab delimeted closure table file to the directory of choice.
    /// </summary>
    /// <param name="outputFile"></param>
    public void CreateClosureTable(String outputFile)
    {
      List<long[]> outputSet = new List<long[]>();
      long currentSource;
      long currentDestination;
      using (StreamWriter sWriter = new StreamWriter(outputFile))
      {
        foreach (SnomedQueryConcept concept in this.snomedConcepts.Values)
        {
          currentSource = concept.ConceptId;
          SnomedQueryConcept[] conceptArray = this.FindDecendents(concept.ConceptId);
          foreach (SnomedQueryConcept descendant in conceptArray)
          {
            currentDestination = descendant.ConceptId;
            outputSet.Add(new long[] { currentSource, currentDestination });
          }
        }
        foreach (long[] pair in outputSet)
        {
          sWriter.WriteLine(String.Format("{0}\t{1}", pair[0], pair[1]));
        }
      }
    }
  }
}